source/common/http/filter_manager.cc

```diff
@@ -1225,6 +1299,50 @@ void FilterManager::maybeEndEncode(bool end_stream) {
   }
 }
 
+void FilterManager::rewriteResponse() {
+  auto response_headers = filter_manager_callbacks_.responseHeaders();
+  ASSERT(response_headers.ptr() != nullptr);
+
+  std::string rewritten_body{};
+  absl::string_view rewritten_content_type{};
+  Http::Code rewritten_code{static_cast<Http::Code>(Utility::getResponseStatus(*response_headers))};
+
+  // Start with the local reply body, if we have it.
+  if (local_reply_data_) {
+    rewritten_body = local_reply_data_->body_text_;
+  }
+
+  ENVOY_STREAM_LOG(trace, "rewriteResponse: calling local_reply_.rewrite with body=\"{}\", code={}",
+                   *this, rewritten_body, rewritten_code);
+  const bool did_rewrite =
+      local_reply_.rewrite(filter_manager_callbacks_.requestHeaders().ptr(), *response_headers,
+                           stream_info_, rewritten_code, rewritten_body, rewritten_content_type);
+  ENVOY_STREAM_LOG(
+      trace, "rewriteResponse: local_reply_.rewrite returned body=\"{}\", content_type={}, code={}",
+      *this, rewritten_body, rewritten_content_type, rewritten_code);
+
+  if (local_reply_data_ && local_reply_data_->is_grpc_) {
+    // Send a trailers-only grpc response
+    Utility::toGrpcTrailersOnlyResponse(*response_headers, rewritten_code, rewritten_body,
+                                        local_reply_data_->grpc_status_,
+                                        local_reply_data_->is_head_request_);
+    // We're sending a trailers-only response with no body. Make sure there's no buffered
+    // response data nor a content length header.
+    buffered_response_data_ = nullptr;
+    return;
+  }
+
+  buffered_response_data_ =
```

Comment:
I decided to use buffered_response_data_ since the filter chain uses this during addEncodedData / modifyEncodingBuffer, and it seemed appropriate. Is this reasonable?